plugins/servlet3/src/main/java/org/apache/struts2/servlet3/async/Servlet3AsyncManager.java

```diff
@@ -0,0 +1,149 @@
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.struts2.servlet3.async;
+
+import com.opensymphony.xwork2.AsyncManager;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.struts2.ServletActionContext;
+
+import javax.servlet.AsyncContext;
+import javax.servlet.AsyncEvent;
+import javax.servlet.AsyncListener;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Implements {@link AsyncManager} to add support for invoke async actions via Servlet 3's API.
+ *
+ * @since 2.5.14
+ */
+public class Servlet3AsyncManager implements AsyncManager, AsyncListener {
+    private static final Logger LOG = LogManager.getLogger(Servlet3AsyncManager.class);
+    private static final AtomicInteger threadCount = new AtomicInteger(0);
+
+    private AsyncContext asyncContext;
+    private boolean asyncActionStarted;
+    private Boolean asyncCompleted;
+    private Object asyncActionResult;
+
+    @Override
+    public void invokeAsyncAction(final Callable asyncAction) {
+        if (asyncActionStarted) {
+            return;
+        }
+
+        Long timeout = null;
+        Executor executor = null;
+        if (asyncAction instanceof AsyncAction) {
+            AsyncAction customAsyncAction = (AsyncAction) asyncAction;
+            timeout = customAsyncAction.getTimeout();
+            executor = customAsyncAction.getExecutor();
+        }
+
+        HttpServletRequest req = ServletActionContext.getRequest();
+        asyncActionResult = null;
+        asyncCompleted = false;
+
+        if (asyncContext == null || !req.isAsyncStarted()) {
+            asyncContext = req.startAsync(req, ServletActionContext.getResponse());
+            asyncContext.addListener(this);
+            if (timeout != null) {
+                asyncContext.setTimeout(timeout);
+            }
+        }
+        asyncActionStarted = true;
+        LOG.debug("Async processing started for " + asyncContext);
+
+        final Runnable task = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    setAsyncActionResultAndDispatch(asyncAction.call());
+                } catch (Throwable e) {
+                    setAsyncActionResultAndDispatch(e);
+                }
+            }
+        };
+        if (executor != null) {
+            executor.execute(task);
+        } else {
+            final Thread thread = new Thread(new Runnable() {
```

Comment:
If apps enter this else branch things are likely slow. So we should provide defaults to avoid this. I cannot find where `executor` is created. Could we initialize it in `AsyncAction`?